collectors/mesos/agent/agent.go

```diff
@@ -53,6 +54,53 @@ type Collector struct {
 	metricsChan chan producers.MetricsMessage
 	nodeInfo    collectors.NodeInfo
 	timestamp   int64
+
+// ContainerTaskRels defines the relationship between containers and tasks.
+type ContainerTaskRels struct {
+	sync.Mutex
+	rels map[string]*TaskInfo
+}
+
+// Get is a utility method which handles the mutex lock and abstracts the inner
+// map in ContainerTaskRels away. If no task info is available for the supplied
+// key, returns nil.
+func (ctr *ContainerTaskRels) Get(key string) *TaskInfo {
+	ctr.Lock()
+	defer ctr.Unlock()
+	if ctr.rels != nil {
+		return ctr.rels[key]
+	}
+	// ContainerTaskRels was not yet available
+	return nil
+}
+
+// Set adds or updates an entry to ContainerTaskRels and, if necessary,
+// initiates the inner map. It is only currently used in tests.
+func (ctr *ContainerTaskRels) Set(key string, info *TaskInfo) {
+	ctr.Lock()
+	if ctr.rels == nil {
+		ctr.rels = map[string]*TaskInfo{}
+	}
+	ctr.rels[key] = info
+	ctr.Unlock()
+}
+
+// update denormalizes the (deeply nested) /state map from the local mesos
+// agent to a list of tasks mapped to container IDs.
+func (ctr *ContainerTaskRels) update(as agentState) {
+	rels := map[string]*TaskInfo{}
+	for _, f := range as.Frameworks {
+		for _, e := range f.Executors {
+			for _, t := range e.Tasks {
+				for _, s := range t.Statuses {
+					rels[s.ContainerStatusInfo.ID.Value] = &t
```

Comment:
I absolutely follow the logic of what you're saying, but I'm a little puzzled because I can't force this bug to emerge. I am running four identical applications on the same node, each called emitter-N where N is in the range 1-4. 

It seems to me that if this bug exists, the metrics results for all four containers should have the same name and task ID. But their task IDs remain distinct, correct and consistent.

Reply:
Since the loop goes first over executors, and then tasks, this bug will not appear if there is only ever one task per executor, I think.  Is it possible that an executor can ever have more than one task?